### PR TITLE
Add support for Protobuf deprecated option

### LINKED ISSUE #3784: Implement option deprecated = true
It would be nice to implement this option for both services and rpc methods. Currently we just ignore it.

### PR DESCRIPTION
**Pull request by pepone**
Fix #3784

### CODE REVIEW

**Review by externl — APPROVED** [review]

File: src/IceRpc.Protobuf.Generators/Internal/Emitter.cs [hunk 1]
```diff
@@ -82,7 +82,19 @@ request.Operation switch
             generatedClasses.Add(container);
         }
 
-        string generated = "using IceRpc.Protobuf;\n\n";
+        string generated = @$"
+// <auto-generated/>
```
Comment (bernardnormier) [inline src/IceRpc.Protobuf.Generators/Internal/Emitter.cs:86]: Should we do the same for IceRpc.Slice.Generators?

File: tools/IceRpc.ProtocGen/ClientGenerator.cs [hunk 2]
```diff
@@ -18,20 +18,32 @@ internal static string GenerateInterface(ServiceDescriptor service)
             string returnType = method.OutputType.GetType(scope, method.IsServerStreaming);
             string methodName = $"{method.Name.ToPascalCase()}Async";
 
-            methods += $@"
+            if (method.GetOptions()?.Deprecated ?? false)
+            {
+                methods += @$"
+    [global::System.Obsolete]".Trim();
```
Comment (bernardnormier) [inline tools/IceRpc.ProtocGen/ClientGenerator.cs:24]: I don't understand the call to `Trim`.

File: tools/IceRpc.ProtocGen/ClientGenerator.cs [hunk 3]
```diff
@@ -18,20 +18,32 @@ internal static string GenerateInterface(ServiceDescriptor service)
...
         }
-        return @$"
+
+        string clientInterface = @$"
 /// <remarks>protoc-gen-icerpc-csharp generated this client-side interface from Protobuf service <c>{service.FullName}</c>.
-/// It's implemented by <c>{service.Name.ToPascalCase()}Client</c></remarks>
+/// It's implemented by <c>{service.Name.ToPascalCase()}Client</c></remarks>";
+        if (service.GetOptions()?.Deprecated ?? false)
+        {
+            clientInterface += "[global::System.Obsolete]";
```
Comment (bernardnormier) [inline tools/IceRpc.ProtocGen/ClientGenerator.cs:39]: It seems we're missing a newline before this attribute.

File: tools/IceRpc.ProtocGen/ClientGenerator.cs [hunk 4]
```diff
@@ -60,7 +72,12 @@ internal static string GenerateImplementation(ServiceDescriptor service)
                 (true, true) => "InvokeBidiStreamingAsync",
             };
 
-            methods += $@"
+            if (method.GetOptions()?.Deprecated ?? false)
+            {
+                methods += @"
+    [global::System.Obsolete]".Trim();
```
Comment (bernardnormier) [inline tools/IceRpc.ProtocGen/ClientGenerator.cs:78]: Why Trim?

File: tools/IceRpc.ProtocGen/ClientGenerator.cs [hunk 5]
```diff
@@ -79,10 +96,15 @@ internal static string GenerateImplementation(ServiceDescriptor service)
...
 
-        return @$"
+        string clientImplementation = @$"
 /// <summary>Makes invocations on a remote IceRPC service. This remote service must implement Protobuf service
 /// <c>{service.FullName}</c>.</summary>
-/// <remarks>protoc-gen-icerpc-csharp generated this record struct from Protobuf service <c>{service.FullName}</c>.</remarks>
+/// <remarks>protoc-gen-icerpc-csharp generated this record struct from Protobuf service <c>{service.FullName}</c>.</remarks>";
+        if (service.GetOptions()?.Deprecated ?? false)
+        {
+            clientImplementation += "[global::System.Obsolete]";
```
Comment (bernardnormier) [inline tools/IceRpc.ProtocGen/ClientGenerator.cs:105]: Missing newline?

File: tools/IceRpc.ProtocGen/ServiceGenerator.cs [hunk 6]
```diff
@@ -37,24 +37,36 @@ public static string GenerateInterface(ServiceDescriptor service)
     /// <param name=""{inputParam}"">{inputParamDocComment}</param>
     /// <param name=""features"">The dispatch features.</param>
     /// <param name=""cancellationToken"">A cancellation token that receives the cancellation requests.</param>
-    /// <returns>{returnTypeDocComment}</returns>
+    /// <returns>{returnTypeDocComment}</returns>";
+            if (method.GetOptions()?.Deprecated ?? false)
+            {
+                methods += @"
+    [global::System.Obsolete]".Trim();
```
Comment (bernardnormier) [inline tools/IceRpc.ProtocGen/ServiceGenerator.cs:44]: Same as above.

**Review by bernardnormier — APPROVED** [review]

File: src/IceRpc.Protobuf.Generators/Internal/Emitter.cs (hunk 1 above)
Comment (pepone) [reply src/IceRpc.Protobuf.Generators/Internal/Emitter.cs:86]: Yes, I created an issue for this.

File: tools/IceRpc.ProtocGen/ClientGenerator.cs (hunk 2 above)
Comment (pepone) [reply tools/IceRpc.ProtocGen/ClientGenerator.cs:24]: It is not required.

File: tools/IceRpc.ProtocGen/ClientGenerator.cs (hunk 3 above)
Comment (pepone) [reply tools/IceRpc.ProtocGen/ClientGenerator.cs]: Added new line

**Pull request merged**